### ISSUE TITLE
Ltac2: rewrite with local hypothesis does not work

### ISSUE THREAD

**Issue by MSoegtropIMC** [rocq-prover/rocq#17778]
#### Description of the problem

In the following example (from CUDW2023) rewriting with a local hypothesis does not work using Ltac2 rewrite:

```
Require Import Reals.
Require Import QArith.
Require Import Ltac2.Ltac2.

Lemma simplify_R_Opp: forall (r : R) (q : Q) (eq : r = Q2R q),
  (- r = Q2R (-q))%R.
Proof.
  intros r q eq.
  Fail rewrite eq.  (* rewriting with a local hypothesis does not seem to work with Ltac2 rewrite *)
  ltac1:(rewrite eq). (* using ltac1 rewrite does work *)
  rewrite Qreals.Q2R_opp. (* Using global lemmas also works with Ltac2 rewrite *)
  reflexivity.
Qed.
```

#### Coq Version

```
The Coq Proof Assistant, version 8.17.0
compiled with OCaml 4.14.1
```

**Comment by ppedrot** [comment]
You have to write `rewrite &eq`, otherwise you get the global constant `eq`, hence the error. It's not related to rewrite.

**Comment by ppedrot** [comment]
This behaviour is even documented [here](https://coq.inria.fr/refman/proof-engine/ltac2.html#strict-vs-non-strict-mode):
> In non-strict mode, any simple identifier appearing in a term quotation which is not bound in the global environment is turned into a dynamic reference to a hypothesis.

Since `eq` is globally bound, it's resolved there.

**Comment by MSoegtropIMC** [comment]
Hmm - I thought the rules were different in proof mode and when writing an Ltac2 function. Wasn't it so that in proof mode local symbols take priority over global symbols?

**Comment by ppedrot** [comment]
No, it's a conscious design choice. We emulate the "non-strict" Ltac1 behaviour when there is no ambiguity, but for stability of code factorization we explicitly favour the global constant over the local one otherwise. This makes it easy to move code between scripts and definitions without changing observable behaviour, and it also makes it very easy to be compatible with both strict and non-strict mode (just put ampersand everywhere instead of implicitly relying on the dynamic interpretation). Ideally we shouldn't even have the non-strict mode (it's a modularity can of worms) but it's there for backwards compatibility purposes with Ltac1.

**Comment by MSoegtropIMC** [comment]
I see - makes sense. Maybe it would help to put this reasoning in the reference manual.

Please fell free to either close or convert it to a doc issue - as you see fit.
